scipy/sparse/linalg/_dsolve/linsolve.py

```diff
@@ -651,93 +649,84 @@ def spsolve_triangular(A, b, lower=True, overwrite_A=False, overwrite_b=False,
     Examples
     --------
     >>> import numpy as np
-    >>> from scipy.sparse import csr_matrix
+    >>> from scipy.sparse import csc_array
     >>> from scipy.sparse.linalg import spsolve_triangular
-    >>> A = csr_matrix([[3, 0, 0], [1, -1, 0], [2, 0, 1]], dtype=float)
+    >>> A = csc_array([[3, 0, 0], [1, -1, 0], [2, 0, 1]], dtype=float)
     >>> B = np.array([[2, 0], [-1, 0], [2, 0]], dtype=float)
     >>> x = spsolve_triangular(A, B)
     >>> np.allclose(A.dot(x), B)
     True
     """
 
     if is_pydata_spmatrix(A):
-        A = A.to_scipy_sparse().tocsr()
+        A = A.to_scipy_sparse().tocsc()
 
-    # Check the input for correct type and format.
-    if not (issparse(A) and A.format == "csr"):
-        warn('CSR matrix format is required. Converting to CSR matrix.',
+    if not (issparse(A) and A.format == "csc"): 
+        warn('CSC matrix format is required. Converting to CSC matrix.',
```

Comment:
This change is tricky, since it forces a change on downstream users of `spsolve_triangular`. That makes this a PR with a backwards compatibility impact, where to avoid the warning a user must choose CSC or CSR depending on SciPy version. How necessary is it to make the new implementation use CSC?

Reply:
Thank you for taking a look! I have added direct support for CSR matrices in d645413d, which required some extension of the SuperLU gstrs wrapper in 9dce866d. 

In short, the matrices L and U must be passed to the SuperLU  gstrs functions in CSC-like format. gstrs takes an argument that allows to solve the transposed system instead. By the above changes, if we get a CSR matrix in `spsolve_triangular`, we transpose it to get a CSC matrix, and then have SuperLU solve the transposed system.

Reply:
Besides the CSR->CSC conversion that has been removed now, there is another conversion to LIL and back if `unit_diagonal=True`, in order to be able to set the diagonal entries. I hope that this is ok given the overall performance improvement, but please let me know if a warning should be raised or added to the documentation.

Reply:
Thanks! I'll resolve this comment, looks good to me now. The `unit_diagonal=True` seems okay, that's not the default, looks pretty niche at first sight, and I did a code search to confirm that its usage is quite rare.